### ISSUE TITLE
why data-preprocessing is different between training and inference?

### ISSUE THREAD

**Issue by Fred199683** [ultralytics/yolov5#3054]
## ❔Question
Thanks for ur repo, i have read ur code and have some questions about the data preprocess.
if i have an image of size(1280x1000), and the model requires 640. if in inference, the image will be resized to 640x500, and pad to 640x512. But if in training, if i don't use mosaic, so the image will just resized to 640x640(this will stretch the image).
I think this will hurt the model performance, because the images in training is all stretched, but in inference is not.

plus: if  i use mosaic in training, it will not use the letterbox function, so it will not have the problem above?

hope someone can help me with this! thx!
## Additional context
`def letterbox(img, new_shape=(640, 640), color=(114, 114, 114), auto=True, scaleFill=False, scaleup=True, stride=32):
    # Resize and pad image while meeting stride-multiple constraints
    shape = img.shape[:2]  # current shape [height, width]
    if isinstance(new_shape, int):
        new_shape = (new_shape, new_shape)

    # Scale ratio (new / old)
    r = min(new_shape[0] / shape[0], new_shape[1] / shape[1])
    if not scaleup:  # only scale down, do not scale up (for better test mAP)
        r = min(r, 1.0)

    # Compute padding
    ratio = r, r  # width, height ratios
    new_unpad = int(round(shape[1] * r)), int(round(shape[0] * r))
    dw, dh = new_shape[1] - new_unpad[0], new_shape[0] - new_unpad[1]  # wh padding
    if auto:  # minimum rectangle
        dw, dh = np.mod(dw, stride), np.mod(dh, stride)  # wh padding
    elif scaleFill:  # stretch
        dw, dh = 0.0, 0.0
        new_unpad = (new_shape[1], new_shape[0])
        ratio = new_shape[1] / shape[1], new_shape[0] / shape[0]  # width, height ratios

    dw /= 2  # divide padding into 2 sides
    dh /= 2

    if shape[::-1] != new_unpad:  # resize
        img = cv2.resize(img, new_unpad, interpolation=cv2.INTER_LINEAR)
    top, bottom = int(round(dh - 0.1)), int(round(dh + 0.1))
    left, right = int(round(dw - 0.1)), int(round(dw + 0.1))
    img = cv2.copyMakeBorder(img, top, bottom, left, right, cv2.BORDER_CONSTANT, value=color)  # add border
    return img, ratio, (dw, dh)`
`        if mosaic:
            # Load mosaic
            img, labels = load_mosaic(self, index)
            shapes = None

            # MixUp https://arxiv.org/pdf/1710.09412.pdf
            if random.random() < hyp['mixup']:
                img2, labels2 = load_mosaic(self, random.randint(0, self.n - 1))
                r = np.random.beta(8.0, 8.0)  # mixup ratio, alpha=beta=8.0
                img = (img * r + img2 * (1 - r)).astype(np.uint8)
                labels = np.concatenate((labels, labels2), 0)

        else:
            # Load image
            img, (h0, w0), (h, w) = load_image(self, index)

            # Letterbox
            shape = self.batch_shapes[self.batch[index]] if self.rect else self.img_size  # final letterboxed shape
            img, ratio, pad = letterbox(img, shape, auto=False, scaleup=self.augment)
            shapes = (h0, w0), ((h / h0, w / w0), pad)  # for COCO mAP rescaling

            labels = self.labels[index].copy()
            if labels.size:  # normalized xywh to pixel xyxy format
                labels[:, 1:] = xywhn2xyxy(labels[:, 1:], ratio[0] * w, ratio[1] * h, padw=pad[0], padh=pad[1])`

**Comment by glenn-jocher** [comment]
👋 Hello, thank you for asking about the differences between [train.py](https://github.com/ultralytics/yolov5/blob/master/train.py), [detect.py](https://github.com/ultralytics/yolov5/blob/master/detect.py) and [test.py](https://github.com/ultralytics/yolov5/blob/master/test.py) in YOLOv5.

These 3 files are designed for different purposes and utilize different dataloaders with different settings. train.py dataloaders are designed for a speed-accuracy compromise, test.py is designed to obtain the best mAP on a validation dataset, and detect.py is designed for best real-world inference results. A few important aspects of each:

## train.py

- **trainloader** LoadImagesAndLabels(): designed to load train dataset images and labels. Augmentation capable and enabled. https://github.com/ultralytics/yolov5/blob/fca5e2a48fb526b57bda0c66be6b7ac1aaa8d83d/train.py#L188-L192
- **testloader** LoadImagesAndLabels(): designed to load val dataset images and labels. Augmentation capable but disabled. https://github.com/ultralytics/yolov5/blob/fca5e2a48fb526b57bda0c66be6b7ac1aaa8d83d/train.py#L199-L202
- image size: 640
- [confidence threshold](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/test.py#L291): 0.001
- [iou threshold](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/test.py#L292): 0.6
- [multi-label](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/test.py#L120): True
- [padding](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/test.py#L83-L90): None

## test.py

- **dataloader** LoadImagesAndLabels(): designed to load train, val, test dataset images and labels. Augmentation capable but disabled. https://github.com/ultralytics/yolov5/blob/fca5e2a48fb526b57bda0c66be6b7ac1aaa8d83d/test.py#L89-L90
- image size: 640
- [confidence threshold](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/test.py#L291): 0.001
- [iou threshold](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/test.py#L292): 0.6
- [multi-label](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/test.py#L120): True
- [padding](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/test.py#L83-L90): 0.5 * maximum stride

## detect.py

- **dataloaders** (multiple): designed for loading multiple types of media (images, videos, globs, directories, streams). https://github.com/ultralytics/yolov5/blob/fca5e2a48fb526b57bda0c66be6b7ac1aaa8d83d/detect.py#L46-L53
- image size: 640
- [confidence threshold](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/detect.py#L152): 0.25 
- [iou threshold](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/detect.py#L153): 0.45
- [multi-label](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/utils/general.py#L415): False
- [padding](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/detect.py#L45-L54): None

## YOLOv5 PyTorch Hub

- autoShape() class used for image loading, preprocessing, inference and NMS. For more info see [YOLOv5 PyTorch Hub Tutorial](https://docs.ultralytics.com/yolov5/tutorials/pytorch_hub_model_loading) https://github.com/ultralytics/yolov5/blob/fca5e2a48fb526b57bda0c66be6b7ac1aaa8d83d/models/common.py#L225-L250
